sdk/core/core/src/az_http_response.c

```diff
@@ -258,3 +259,15 @@ AZ_NODISCARD az_result az_http_response_get_body(az_http_response* response, az_
   response->_internal.parser.next_kind = _az_HTTP_RESPONSE_KIND_EOF;
   return AZ_OK;
 }
+
+void _az_http_response_reset(az_http_response* http_response)
+{
+  // never fails, discard the result
```

Comment:
If `az_http_response_init` can never fail, why does it return an `az_result`? Or do we suspect it could fail in the future as the API/impl evolves?

Reply:
init functions return az_result because that is consistent and, over time, the init could fail. Although, this is probably not possible in reality because it's unlike that customers will check this return value in their code.
az_span_init is an exception; we could rename it if we really want to: az_span(...) or az_span_make(...)?

Reply:
> this is probably not possible in reality because it's unlike that customers will check this return value in their code.

Hmm, in that case, why are we forcing customers to check for it with `AZ_NODISCARD` (FWIW, I think if we return `az_result`, we **should** keep the no discard)?

If the pattern is that `_init` functions are meant to return `az_result` and they could fail (potentially in the future), we need to make sure callers always check the result and either bubble up the error or take action.

If some `_init` functions will never fail, maybe we shouldn't confuse the user by returning an `az_result` that is always `AZ_OK`. The callers can then end up writing code like what we have in this PR, where they ignore the result, which goes against the coding pattern we want to establish.

> az_span_init is an exception; we could rename it if we really want to: az_span(...) or az_span_make(...)?

Do we have to rename the method away from `init` if it doesn't return `az_result` or can those be independent things? What's the concern with the following:
- Some `_init` functions return an `az_result` that the caller must check because they could fail.
- Other `_init` functions don't fail, and hence return something other than `az_result` (like `az_span`, or void if initializing by some ptr).

If we really want different names, how about the `_ctor` suffix?

Reply:
A function's signature is based on its contract. If the signature says that the function returns an az_result then its contract says it can fail and customers SHOULD ALWAYS check this value (so AZ_NODISCARD is the right thing to have here). And, we are free to return new error codes in the future. 

However, history has shown that sometimes customers ignore the contract. If they learn that a function never returns an error (today), then they may ignore the value (NODISCARD makes this harder). In this case, if we modify the function to return an error, then we end up breaking customer code because they didn't adhere to our contract. In my experience, this sometimes makes it so that we cannot return a new error because it ends up breaking customer code even though it's the customer's fault for not adhering to our contract. If it's a big break, the customer-assumed contract becomes the actual defacto contract and we can't evolve an API.

This has definitely been an issue for Windows in the past but I do not think this will be an issue  for us (partially because we ship source, not DLLs).

I think we can have some _init functions that return az_result and some that don't (az_span_init). But we're setting a contract here for future: functions that do not return an az_result, never can and functions that return an az_result always will.